cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/edge/PropertyEdge.kt

```diff
@@ -54,271 +47,76 @@ import org.slf4j.LoggerFactory
-                        if (obj is PropertyEdge<*>) {
-                            if (outgoing) {
-                                outputCollection.add(obj.end)
-                            } else {
-                                outputCollection.add(obj.start)
-                            }
-                        }
-                    }
-                    return outputCollection
-                } catch (e: InstantiationException) {
-                    log.warn("PropertyEdges could not be unwrapped")
-                } catch (e: IllegalAccessException) {
-                    log.warn("PropertyEdges could not be unwrapped")
-                } catch (e: InvocationTargetException) {
-                    log.warn("PropertyEdges could not be unwrapped")
-                } catch (e: NoSuchMethodException) {
-                    log.warn("PropertyEdges could not be unwrapped")
-                }
-            }
-            return collection
-        }
-
-        /**
-         * @param obj PropertyEdge or collection of property edges that must be unwrapped
-         * @param outgoing direction of the edge
-         * @return node or collection representing target of edge
-         */
-        @JvmStatic
-        fun unwrapPropertyEdge(obj: Any, outgoing: Boolean): Any {
-            if (obj is PropertyEdge<*>) {
-                return if (outgoing) {
-                    obj.end
-                } else {
-                    obj.start
-                }
-            } else if (obj is Collection<*> && obj.isNotEmpty()) {
-                return unwrapPropertyEdgeCollection(obj, outgoing)
-            }
-            return obj
-        }
-
-        /**
-         * Checks if an Object is a PropertyEdge or a collection of PropertyEdges
-         *
-         * @param f Field containing the object
-         * @param obj object that is checked if it is a PropertyEdge
-         * @return true if obj is/contains a PropertyEdge
-         */
-        @JvmStatic
-        fun checkForPropertyEdge(f: Field, obj: Any?): Boolean {
-            if (obj is PropertyEdge<*>) {
-                return true
-            } else if (obj is Collection<*>) {
-                val collectionTypes =
-                    listOf(*(f.genericType as ParameterizedType).actualTypeArguments)
-                for (t in collectionTypes) {
-                    if (t is ParameterizedType) {
-                        return t.rawType == PropertyEdge::class.java
-                    } else if (PropertyEdge::class.java == t) {
-                        return true
-                    }
-                }
-            }
-            return false
+    /**
+     * This special hash-code function takes into account whether the caller put this edge into an
+     * outgoing or incoming edge. This is primarily used by [PropertyEdgeList.hashCode].
+     */
+    fun directionalHashCode(outgoing: Boolean = true): Int {
+        return if (outgoing) {
+            Objects.hash(end)
```

Comment:
We need to exclude the node where we "start" because otherwise we would go into the node when we calculate the hashcode, this would then include the hashcode of the property edge list, which will then loop.

Reply:
But we do use `start` in `hashCode`, isn't this the same problem then?

Reply:
Moved the function outside of property edge because it is only used in edge lists to avoid loops.